File: jinja/javascript/genciconf.js
const fs = require('fs');
const nunjucks = require('nunjucks');

function getDSMList() {
  const settingsGradleFile = "settings.gradle";
  const dsmList = [];
  const marker = "include '";

  // TODO: use another method to throw an error if settings.gradle not found
  const lines = fs.readFileSync(settingsGradleFile, 'utf8').split('\n');
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].includes(marker)) {
      dsmList.push(lines[i].substring(9, lines[i].length - 1));
    }
  }

  return dsmList;
}

function getTriggerSnippet(dsm) {
  const renderedText = nunjucks.renderString(`Build {{ dsm }}:
  stage: Triggers
  trigger:
    include: {{ dsm }}/.gitlab-ci.yml
    strategy: depend
  rules:
    - changes:
        - "**/*"`, { dsm: dsm });

  return renderedText;

}

function genTriggerYaml() {
  const triggerYaml = "triggers.yml";
  const dsmList = getDSMList();

  for (let i = 0; i < dsmList.length; i++) {
    fs.appendFile(triggerYaml, getTriggerSnippet(dsmList[i]) + "\n\n", (err) => {
      if (err) throw err;
    });
  }
}

const main = () => {
  genTriggerYaml();
};

if (require.main === module) {
  main();
}

File: jinja/lua/genciconf.lua
local template = require "resty.template"

function getDSMList()
    local settingsGradleFile = "settings.gradle"
    local dsmList = {}
    local marker = "include '"

    local file, err = io.open(settingsGradleFile, "r")
    if not file then error("File not found: " .. settingsGradleFile) end

    for line in file:lines() do
        if string.find(line, marker) then
            table.insert(dsmList, string.sub(line, 10, string.len(line) - 1))
        end
    end
    file:close()

    return dsmList
end

function getTriggerSnippet(dsm)
    local renderedText = template.process([[
Build {{ dsm }}:
  stage: Triggers
  trigger:
    include: {{ dsm }}/.gitlab-ci.yml
    strategy: depend
  rules:
    - changes:
        - "**/*"
]], {dsm = dsm})

    return renderedText

end

function genTriggerYaml()
    local triggerYaml = "triggers.yml"
    local dsmList = getDSMList()
    for _, dsm in ipairs(dsmList) do
        local file = io.open(triggerYaml, "a")
        do
            file:write(getTriggerSnippet(dsm))
            file:write("\n")
            file:close()
        end
    end
end

function main() genTriggerYaml() end

main()
